In [ ]:
import numpy as np
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [ ]:
import pandas as pd
data = pd.read_csv("/content/drive/My Drive/train.csv")

In [ ]:
data

Id  minutes_past  ...  Kdp_5x5_90th  Expected
0               1             3  ...           NaN  0.254000
1               1            16  ...           NaN  0.254000
2               1            25  ...           NaN  0.254000
3               1            35  ...           NaN  0.254000
4               1            45  ...           NaN  0.254000
...           ...           ...  ...           ...       ...
13765196  1180945            38  ...           NaN  8.636004
13765197  1180945            42  ...           NaN  8.636004
13765198  1180945            47  ...     13.379990  8.636004
13765199  1180945            52  ...           NaN  8.636004
13765200  1180945            57  ...      6.069992  8.636004

[13765201 rows x 24 columns]

In [ ]:
data = data.dropna(axis = 0, subset = ['Ref'])
data = data.dropna(axis = 0, subset = ['RhoHV'])
data.reset_index(drop=True, inplace=True)

In [ ]:
train_id = data[~np.isnan(data.Ref)].Id.unique()
new_data = data[np.in1d(data.Id, train_id)]
del train_id
data = new_data
data.shape

(4069799, 24)

In [ ]:
data.fillna(0.0, inplace=True)
data.reset_index(drop=True, inplace=True)

In [ ]:
data

Id  minutes_past  ...  Kdp_5x5_90th  Expected
0              2             1  ...      1.059998  1.016000
1              2             6  ...      1.409988  1.016000
2              2            11  ...      1.759994  1.016000
3              2            16  ...      1.059998  1.016000
4              2            21  ...      1.759994  1.016000
...          ...           ...  ...           ...       ...
4069794  1180928            57  ...      2.119995  2.794001
4069795  1180935            41  ...      4.889999  0.508000
4069796  1180935            46  ...      4.229996  0.508000
4069797  1180945            47  ...     13.379990  8.636004
4069798  1180945            57  ...      6.069992  8.636004

[4069799 rows x 24 columns]

In [ ]:
data = data.groupby(['Id']).aggregate({'Expected': 'median',
                                            'radardist_km': 'median',
                                            'Ref': 'median',
                                           'Ref_5x5_10th': 'median',
                                           'Ref_5x5_50th': 'median',
                                           'Ref_5x5_90th': 'median',
                                           'RefComposite': 'median',
                                           'RefComposite_5x5_10th': 'median',
                                           'RefComposite_5x5_50th': 'median',
                                           'RefComposite_5x5_90th': 'median',
                                           'RhoHV': 'median',
                                           'RhoHV_5x5_10th': 'median',
                                           'RhoHV_5x5_50th': 'median',
                                           'RhoHV_5x5_90th': 'median',
                                           'Zdr': 'median',
                                           'Zdr_5x5_10th': 'median',
                                           'Zdr_5x5_50th': 'median',
                                           'Zdr_5x5_90th': 'median',
                                           'Kdp': 'median',
                                           'Kdp_5x5_10th': 'median',
                                           'Kdp_5x5_50th': 'median',
                                           'Kdp_5x5_90th': 'median',}
                                           )

In [ ]:
data.reset_index(drop=True, inplace=True)

In [ ]:
data

Expected  radardist_km  ...  Kdp_5x5_50th  Kdp_5x5_90th
0        1.016000           2.0  ...     -0.350006      1.059998
1        4.064002           9.0  ...      0.000000      7.029999
2        3.225002           8.0  ...     -1.050003      4.459992
3        0.010000          10.0  ...      0.529991      7.599991
4        2.540001          12.0  ...     -0.350006      2.119995
...           ...           ...  ...           ...           ...
518527   0.508000           3.0  ...     -0.349998      2.805000
518528  15.494008          15.0  ...     -0.350006      3.529999
518529   2.794001           5.0  ...      0.174995      1.234993
518530   0.508000          14.0  ...     -4.225006      4.559998
518531   8.636004           9.0  ...     -4.015000      9.724991

[518532 rows x 22 columns]

In [ ]:
ydata = data.iloc[:, 0]
ydata

0          1.016000
1          4.064002
2          3.225002
3          0.010000
4          2.540001
            ...    
518527     0.508000
518528    15.494008
518529     2.794001
518530     0.508000
518531     8.636004
Name: Expected, Length: 518532, dtype: float64

In [ ]:
data = data.drop(columns = 'Expected', axis = 1)

In [ ]:
# Среднее значение
mmean = data.mean(axis=0)
# Стандартное отклонение
mstd = data.std(axis=0)
data  -=  mmean
data /= mstd

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(data, ydata, test_size=0.33, random_state=42 )

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Dropout
from tensorflow.keras.layers import BatchNormalization

In [ ]:
#model = Sequential()
#model.add(Dense(128, activation='relu', input_shape=(X_train.shape[1],)))
#model.add(Dense(32, activation='relu'))
#model.add(Dense(1))

In [ ]:
model = Sequential()
model.add(BatchNormalization())
model.add(Dense(1024, activation='relu'))
model.add(Dropout(0.1))
model.add(BatchNormalization())
model.add(Dense(512, activation='relu'))
model.add(Dropout(0.1))
model.add(BatchNormalization())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.1))
model.add(BatchNormalization())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.1))
model.add(BatchNormalization())
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.1))
model.add(Dense(1))

In [ ]:
#print(model.summary())

ValueError: ignored

In [ ]:
model.compile(optimizer='adam', loss='mse', metrics=['mae'])

In [ ]:
model.fit(X_train, y_train, epochs=10, batch_size=1024, validation_split = 0.2, verbose=1)

Epoch 1/10
272/272 [==============================] - 39s 145ms/step - loss: 29521.7441 - mae: 24.0473 - val_loss: 2105786.5000 - val_mae: 35.0221
Epoch 2/10
272/272 [==============================] - 40s 147ms/step - loss: 28708.0020 - mae: 23.7501 - val_loss: 3476425.2500 - val_mae: 44.2049
Epoch 3/10
272/272 [==============================] - 40s 147ms/step - loss: 28406.7070 - mae: 23.8864 - val_loss: 5594083.0000 - val_mae: 39.4186
Epoch 4/10
272/272 [==============================] - 40s 148ms/step - loss: 28346.4727 - mae: 23.7643 - val_loss: 31736314.0000 - val_mae: 54.6878
Epoch 5/10
272/272 [==============================] - 40s 148ms/step - loss: 28023.3574 - mae: 23.6349 - val_loss: 1090578.6250 - val_mae: 32.0372
Epoch 6/10
272/272 [==============================] - 41s 149ms/step - loss: 28776.2227 - mae: 24.1953 - val_loss: 10519416.0000 - val_mae: 46.9851
Epoch 7/10
272/272 [==============================] - 41s 150ms/step - loss: 28545.8066 - mae: 23.6862 - val_loss: 1

In [ ]:
y_test.reset_index(drop=True, inplace=True)

In [ ]:
pred = model.predict(X_test)

In [ ]:
print("Предсказанное:", pred[1], ", правильное:", y_test[1])

Предсказанное: [8.995802] , правильное: 7.900004400000001


In [ ]:
pred = pred.flatten()
np.corrcoef(pred, y_test)

array([[1.       , 0.0899588],
       [0.0899588, 1.       ]])

In [ ]:
СС_tuner = np.corrcoef(pred, y_test)
СС_tuner = СС_tuner[0][1]
print(f'Коэффициаент корреляция с истинными данными: {СС_tuner}')

Коэффициаент корреляция с истинными данными: 0.0899588042667035


#Коэффициаент корреляция с истинными данными: 0.09888404175803006
#Коэффициаент корреляция с истинными данными: 0.22731226143349065
#Коэффициаент корреляция с истинными данными: 0.06337660802437357
#Коэффициаент корреляция с истинными данными: 0.0899588042667035


In [ ]:
#mmae = 0
#for i in np.arange(0, len(y_test)):
#print("Предсказанная стоимость:", pred[i], ", правильная стоимость:", y_test[i], 'разница: ', np.abs(pred[i] - y_test[i]))
#mmae = np.abs(pred[i] - y_test[i]) + mmae
#final_mae = mmae/292
#final_mae

IndentationError: ignored

In [ ]:
#from matplotlib import pyplot as plt
#y_test = y_test.reset_index(drop=True)

#x = np.arange(0, len(y_test))

#pred = np.round(pred, 0)

#pred = np.reshape(pred, (pred.shape[0]))

#plt.bar(x, y_test, label = 'True')
#plt.bar(x, predicted_y_tuner, label = 'Pred')
#plt.legend()

NameError: ignored

Error in callback <function flush_figures at 0x7f2fbf17a320> (for post_execute):


KeyboardInterrupt: ignored